citra-qt: Add Input Setting dialog

chinhodado opened the pull request:
![image](https://cloud.githubusercontent.com/assets/1388219/7902772/c9e39690-0792-11e5-9343-f96d1ba85b36.png)

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+    QPushButton *resetButton = this->ui.buttonBox->button(QDialogButtonBox::Reset);
+    connect(resetButton, SIGNAL(clicked()), this, SLOT(OnResetClicked()));
+
+    QPushButton *okButton = this->ui.buttonBox->button(QDialogButtonBox::Ok);
```
"ok" and "cancel" should be handled via QDialogButtonBox instead.

neobrain commented on src/citra_qt/KeyPressEater.cpp:
```diff
@@ -0,0 +1,80 @@
+#include <KeyPressEater.h>
```
no license header

neobrain commented on src/citra_qt/KeyPressEater.h:
```diff
@@ -0,0 +1,10 @@
+public:
+    KeyPressEater(QObject* parent);
+    bool eventFilter(QObject *obj, QEvent *event);
+};
```
missing newline at end of file

neobrain commented on src/citra_qt/KeyPressEater.h:
```diff
@@ -0,0 +1,10 @@
+#include <QObject>
```
no license

neobrain commented on src/citra_qt/KeyPressEater.h:
```diff
@@ -0,0 +1,10 @@
+#include <QObject>
+#include <QEvent>
```
Forward declare QEvent instead to minimize compile times.

neobrain commented on src/citra_qt/KeyPressEater.h:
```diff
@@ -0,0 +1,10 @@
+
+/// The event filter used for LineEdits to handle key presses
+class KeyPressEater : public QObject {
+    Q_OBJECT
```
space after Q_OBJECT

neobrain commented on src/citra_qt/KeyPressEater.h:
```diff
@@ -0,0 +1,10 @@
+    Q_OBJECT
+public:
+    KeyPressEater(QObject* parent);
+    bool eventFilter(QObject *obj, QEvent *event);
```
+ `override` ?

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+#include <QSettings>
+#include <QtWidgets>
+#include "inputs.h"
+#include "KeyPressEater.h"
```
include common header first, then core, then the current module. Separate the modules via blank lines

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+
+#include <QKeySequence>
+#include <QSettings>
+#include <QtWidgets>
```
Don't blindly include all of QtWidgets. include the proper headers instead.

neobrain replied:
Also uh, QtWidgets is not Qt4 compatible, which makes the travis build fail (and mine).

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+    // initialize the edit_input_map
+    if (!edit_input_map_initialized) {
+        GInputsDialog::edit_input_map = std::map<QString, Button> {
+            { this->ui.lineEdit_A->objectName(), Button::A},
```
missing space before `}`

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    // set up event handlers for the buttons
+    QPushButton *resetButton = this->ui.buttonBox->button(QDialogButtonBox::Reset);
```
Star should be at the type, not at the variable name.

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+    QPushButton *okButton = this->ui.buttonBox->button(QDialogButtonBox::Ok);
+    connect(okButton, SIGNAL(clicked()), this, SLOT(OnOkClicked()));
+
+    // create a copy of the current settings
```
Why? Comment doesn't explain this.

chinhodado replied:
So that we have a copy of the current button settings as the data model for the user to tinker with behind the scene. 

Of course, another approach is to parse the LineEdits on save to get the settings, but this way we don't need to worry about the parsing logic.

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+
+    // initialize the edit_input_map
+    if (!edit_input_map_initialized) {
+        GInputsDialog::edit_input_map = std::map<QString, Button> {
```
Why don't you instead store a map of QLineEdit pointers to Button?

chinhodado replied:
I was afraid that after the dialog is destroyed, the pointers to the QLineEdits are no longer valid. I will check again though.

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+
+    // setup event filters for the LineEdits
+    KeyPressEater* keyPressEater = new KeyPressEater(this);
+    this->ui.lineEdit_A->installEventFilter(keyPressEater);
```
If you change the line edit map to a map from QLineEdit pointers, you can simply use a range-based for loop to automize this.

neobrain commented on src/citra_qt/inputs.h:
```diff
@@ -0,0 +1,46 @@
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
```
needs more blank lines for separation.

chinhodado replied:
Shouldn't one blank line be enough?

neobrain replied:
Yeah, I think the older version didn't have one, it's fine now either way ;)

neobrain commented on src/citra_qt/inputs.ui:
```diff
@@ -0,0 +1,594 @@
+    <enum>Qt::Horizontal</enum>
+   </property>
+   <property name="standardButtons">
+    <set>QDialogButtonBox::Cancel|QDialogButtonBox::Ok|QDialogButtonBox::Reset</set>
```
Should be `QDialogButtonBox::RestoreDefaults` instead. Reset suggests resetting the settings to what they were upon opening the dialog, while RestoreDefaults actually resets to defaults.

neobrain commented on src/citra_qt/main.ui:
```diff
@@ -141,6 +142,14 @@
+    <string>Configure &amp;Inputs</string>
+   </property>
+   <property name="toolTip">
+    <string>Configure Inputs</string>
```
Does it make any sense to set tooltips for menu actions?

neobrain commented on src/common/logging/log.h:
```diff
@@ -32,7 +32,7 @@ typedef u8 ClassType;
  * Specifies the sub-system that generated the log message.
  *
- * @note If you add a new entry here, also add a corresponding one to `ALL_LOG_CLASSES` in log.cpp.
+ * @note If you add a new entry here, also add a corresponding one to `ALL_LOG_CLASSES` in backend.cpp.
```
Please submit such changes as separate PRs, as they're completely irrelevant and just make it harder for me to give my "LGTM".

neobrain commented on src/citra_qt/main.cpp:
```diff
@@ -252,6 +253,11 @@ void GMainWindow::OnOpenHotkeysDialog()
 }
 
+void GMainWindow::OnOpenInputsDialog()
+{
```
`{` should be on previous line

chinhodado replied:
I was following the style used in this file

neobrain replied:
If it's just about little stuff like this, you don't need to follow the Qt style. For the long run it's better to gradually replace the old coding style with the new one, even if that means styles will be mixed for the mean time.

neobrain commented on src/citra_qt/inputs.h:
```diff
@@ -0,0 +1,46 @@
+    /// Display the button settings on the LineEdits from the given values
+    void displayButtonSettings(Settings::Values values);
+
+private slots:
```
Define private slots before private members, please.

chinhodado replied:
I was following the order like in main.h.

neobrain commented on src/citra_qt/inputs.h:
```diff
@@ -0,0 +1,46 @@
+};
+
+/// The input configuration dialog
+class GInputsDialog : public QDialog {
```
If it's not too complicated, can you change the widget from a dialog to a plain widget, and instead spawn the dialog and set GInputsDialog as its main widget? This will make the (eventually) upcoming config dialog merger a lot easier.

neobrain commented on src/citra_qt/inputs.h:
```diff
@@ -0,0 +1,46 @@
+#include "ui_inputs.h"
+
+/// An enum for the buttons on the 3DS
+enum Button {
```
Don't we have an enum like this somewhere in core already?

chinhodado replied:
AFAIK, no.

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+#include "config.h"
+#include "common/log.h"
+
+std::map<QString, Button> GInputsDialog::edit_input_map;
```
Why are these not class members?

neobrain commented on src/citra_qt/KeyPressEater.cpp:
```diff
@@ -0,0 +1,80 @@
+        int key = key_event->key();
+        switch (button_modified) {
+        case Button::A:
+            GInputsDialog::temp_settings.pad_a_key = key;
```
If it's not too much work, I guess Settings should be restructure to store an array of key values indexed by some Button enum, instead of requiring config code to write everything 10 times.

neobrain commented on src/citra_qt/CMakeLists.txt:
```diff
@@ -36,6 +38,8 @@ set(HEADERS
             util/spinbox.h
             bootmanager.h
             hotkeys.h
+            inputs.h
```
Please create a "config" folder for this.

chinhodado replied:
We already have that folder, and it's currently containing old config code. I'll make another PR later to clean that up and move this stuffs in.

neobrain replied:
Is there any particular issue for not just moving inputs.h there anyway?

neobrain commented on src/citra_qt/CMakeLists.txt:
```diff
@@ -16,6 +16,8 @@ set(SRCS
             bootmanager.cpp
             hotkeys.cpp
+            inputs.cpp
+            KeyPressEater.cpp
```
This should probably go to util/

neobrain commented on src/citra_qt/inputs.ui:
```diff
@@ -0,0 +1,594 @@
+    </rect>
+   </property>
+   <property name="title">
+    <string>D-Pad</string>
```
=>Digital Pad

neobrain replied:
Nevermind, D-Pad is the official name actually.

neobrain commented on src/citra_qt/inputs.ui:
```diff
@@ -0,0 +1,594 @@
+    </rect>
+   </property>
+   <property name="title">
+    <string>Shoulder</string>
```
=> Shoulder Buttons

neobrain replied:
Or actually, going with the official naming it should be "Bumpers".

neobrain commented on src/citra_qt/inputs.ui:
```diff
@@ -0,0 +1,594 @@
+    </rect>
+   </property>
+   <property name="title">
+    <string>Others</string>
```
=> Other Buttons

neobrain commented on src/citra_qt/inputs.ui:
```diff
@@ -0,0 +1,594 @@
+    </rect>
+   </property>
+   <property name="title">
+    <string>Main</string>
```
=> Main Buttons

neobrain commented on src/citra_qt/inputs.ui:
```diff
@@ -0,0 +1,594 @@
+    </rect>
+   </property>
+   <property name="title">
+    <string>C-Pad</string>
```
=> Circle Pad

neobrain commented on src/citra_qt/inputs.ui:
```diff
@@ -0,0 +1,594 @@
+   </rect>
+  </property>
+  <property name="windowTitle">
+   <string>Input setting</string>
```
=> Input Settings

neobrain commented:
Needs some more work, but looking good overall, thanks :)

lioncash commented on src/citra_qt/KeyPressEater.h:
```diff
@@ -0,0 +1,10 @@
+#include <QEvent>
+
+/// The event filter used for LineEdits to handle key presses
+class KeyPressEater : public QObject {
```
This can probably be final if it's not intended to be extended.

lioncash commented on src/citra_qt/inputs.h:
```diff
@@ -0,0 +1,46 @@
+    Ui::inputs ui;
+
+    /// Display the button settings on the LineEdits from the given values
+    void displayButtonSettings(Settings::Values values);
```
Couldn't this be passed by const reference?

lioncash commented on src/citra_qt/inputs.h:
```diff
@@ -0,0 +1,46 @@
+    Q_OBJECT
+
+public:
+    GInputsDialog(QWidget* parent = NULL);
```
nullptr

lioncash commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,155 @@
+    this->ui.lineEdit_sLeft->setText(GInputsDialog::getKeyName(values.pad_sleft_key));
+    this->ui.lineEdit_sRight->setText(GInputsDialog::getKeyName(values.pad_sright_key));
+}
+
```
This can just be:

```cpp
QString GInputsDialog::getKeyName(int key_code)  {
    if (key_code == Qt::Key_Shift)
        return tr("Shift");
    if (key_code == Qt::Key_Control)
        return tr("Ctrl");
    if (key_code == Qt::Key_Alt)
        return tr("Alt");
    if (key_code == Qt::Key_Meta)
        return tr("Meta")

        return QKeySequence(key_code).toString();
}
```

since the returns eliminate the need for an else.

neobrain commented on src/citra_qt/KeyPressEater.h:
```diff
@@ -0,0 +1,10 @@
+#include <QObject>
+#include <QEvent>
+
+/// The event filter used for LineEdits to handle key presses
```
Needs a more elaborate explanation as to what this is exactly intended to do and to be used for.

neobrain commented on src/citra_qt/KeyPressEater.cpp:
```diff
@@ -0,0 +1,80 @@
+
+KeyPressEater::KeyPressEater(QObject* parent) : QObject(parent) {}
+
+bool KeyPressEater::eventFilter(QObject* obj, QEvent* event) {
```
What's the advantage of installing an event filter compared to implementing `QLineEdit::keyPressEvent`? I'd prefer the latter if possible, because currently keyPressEater looks like a general purpose class, but actually is very hardcoded against GInputsDialog (and requires global variables to implement this).

chinhodado commented:
Addressed most comments.

neobrain commented on src/citra_qt/QLineEditKeyConfig.cpp:
```diff
@@ -0,0 +1,72 @@
+    default:
+        break;
+    }
+}
```
no newline at eof

neobrain commented on src/citra_qt/QLineEditKeyConfig.h:
```diff
@@ -0,0 +1,26 @@
+class QKeyEvent;
+
+/// An enum for the buttons on the 3DS
+enum Button {
```
I'm still sure there's something in our core code for this. I'll look it up eventually if you can't find it :)

neobrain commented on src/citra_qt/inputs.cpp:
```diff
@@ -0,0 +1,138 @@
+    ui.setupUi(this);
+
+    // setup the LineEdits
+    this->ui.lineEdit_A->button = Button::A;
```
Keeping around an array (or a std::map) of QLineEdits around instead makes for a lot less code duplication.

neobrain commented on src/citra_qt/inputs.h:
```diff
@@ -0,0 +1,34 @@
+
+private slots:
+    void OnDefaultClicked();
+    void OnOkClicked();
```
Don't call slots by how they are invoked, but by what they actually do, please.

chinhodado replied:
Can you give an example?

chinhodado replied:
Like "OnSaveSettings", "OnRestoreDefaultSettings"?

neobrain replied:
Yeah, just without the "On" prefix ;)

chinhodado replied:
Ok done

neobrain commented on src/citra_qt/inputs.h:
```diff
@@ -0,0 +1,34 @@
+    GInputsDialog(QWidget* parent = nullptr);
+
+    /// Temporary settings used when configuration is changed but not saved yet
+    static Settings::Values temp_settings;
```
Having this as `static` restricts usage of this class to only allow for one instance of the dialog. Surely enough this is good enough for our purposes, but it's bad design regardless since there's no **actual** reason to have this copy be `static`.

neobrain commented on src/citra_qt/CMakeLists.txt:
```diff
@@ -36,6 +38,8 @@ set(HEADERS
             bootmanager.h
             hotkeys.h
+            inputs.h
+            QLineEditKeyConfig.h
```
Please move this to the `utils` folder

neobrain replied:
Actually, nevermind. It's not really a generic utility.

neobrain commented:
Can you update the screenshot?

chinhodado commented:
Updated screenshots and made the changes.

neobrain commented on src/citra_qt/CMakeLists.txt:
```diff
@@ -35,17 +39,18 @@ set(HEADERS
             util/spinbox.h
             bootmanager.h
-            hotkeys.h
+            hotkeys.h            
```
Trailing whitespace

neobrain commented on src/citra_qt/config/QLineEditKeyConfig.cpp:
```diff
@@ -0,0 +1,75 @@
+    this->setText(key_name);
+
+    GInputsDialog* parent_dialog = static_cast<GInputsDialog*>(this->parent()->parent());
+    Settings::Values& temp_settings = parent_dialog->temp_settings;
```
Why not pass a reference to this object in the constructor?

chinhodado replied:
I think this way is fine too.

neobrain replied:
It requires you to use this ridiculous static_cast from this->parent()->parent(), though. In particular that makes an assumption about where this widget is used in, even though there is no reason for that to be the case though - all which this widget needs is access to the Settings object.

chinhodado replied:
Hm right. I'm just hesitant to do it since I'm not involved in constructing these currently (leaving it to Qt Creator with the .ui file) If I have to create them manually I'll also have to position them manually, which kinda defeats the purpose of the ui file.

neobrain replied:
What do you mean with "I'll have to position them manually"? How are the widgets positioned and aligned currently?

chinhodado replied:
It's currently being done in the .ui file (like [here](https://github.com/citra-emu/citra/pull/531/files#diff-6ec8e4f84147ccb1d8597bceaace68c7R44)). I don't want to do that positioning in the code.

neobrain replied:
I.e. you are hardcoding all of the positions? Positioning is usually left to the UI by using layouts instead. Hardcoding positions is just crying for trouble, since the layout will be broken as soon as someone uses a different font family/size than your system.

chinhodado replied:
I think we agreed to leave that for later? Also using UI layouts is a bit hard for this dialog due to the way it's presented (e.g. it's not really a grid)

neobrain replied:
I don't think it makes sense for this to be done in a later PR (I don't recall having agreed on something like that in the first place fwiw).

What issue in particular do you see with UI layouts? As far as I can see, you can just use a VBoxLayout for the main layout, and have the "main buttons" and the Pad groups in a GridLayout (each of which is probably a gridlayout too).

chinhodado commented:
@neobrain any further comments?

chinhodado commented:
Sorry I'm a bit busy atm, I'll get back to this soon.

Qu3ntin commented:
Hey guys I need help how to open citra can someone please help me I really want to contribute but also know how to open and run as shown in the preview

linkmauve commented:
As of #595, we now support New 3DS buttons, including ZL, ZR and the C-stick, those should be added to that dialog.

archshift commented:
What's the status of this?

chinhodado commented:
It is functional, but neobrain disagrees on some design choices that I need to find the time to address.

chinhodado commented:
![image](https://cloud.githubusercontent.com/assets/1388219/7902772/c9e39690-0792-11e5-9343-f96d1ba85b36.png)

Hopefully addressed the rest of @neobrain 's comments.

I changed the look of the dialog since the old look is too hard to achieve with proper layouts, with little benefit.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+    QMetaObject::connectSlotsByName(this);
+
+    setLayout(verticalLayout);
+    setWindowTitle("Input Settings");
```
tr()

neobrain commented on src/citra_qt/config/inputs.h:
```diff
@@ -0,0 +1,59 @@
+    static QString getKeyName(int key_code);
+
+private:
+    QLineEditKeyConfig *lineEdit_A;
```
You can store all of these in an array

neobrain replied:
.. actually, you already do this. uhm..

neobrain commented on src/citra_qt/config/inputs.h:
```diff
@@ -0,0 +1,59 @@
+    QLineEditKeyConfig *lineEdit_Home;
+
+    /// A vector for all the lineEdits above for easy iterating though them
+    std::vector<QLineEditKeyConfig*> lineEdits;
```
Why do you store pointers to the widgets separately from this vector?

Why is this a vector to begin with instead of a plain array?

chinhodado replied:
> Why do you store pointers to the widgets separately from this vector?

Since I need to access them individually later (in `displayButtonSettings()`)

> Why is this a vector to begin with instead of a plain array?

No particular reason, but is there also any reason not to?

neobrain replied:
> No particular reason, but is there also any reason not to?
There is no reason to use the more complex object `vector` in this place.

> Since I need to access them individually later (in displayButtonSettings())
You could instead use an std::map, mapping an enum of keys (we should have one in core already) to widgets. That way you wouldn't need to store the same things twice. If the enum has consecutive values (0, 1, 2, ...), it can even simply be an std::array.

neobrain commented on src/citra_qt/config/inputs.h:
```diff
@@ -0,0 +1,59 @@
+    std::vector<QLineEditKeyConfig*> lineEdits;
+
+    /// Display the button settings on the LineEdits from the given values
+    void displayButtonSettings(const Settings::Values& values);
```
Capital D

neobrain commented on src/citra_qt/main.ui:
```diff
@@ -153,6 +154,11 @@
   </action>
+  <action name="action_Inputs">
+   <property name="text">
+    <string>Configure &amp;Inputs</string>
```
s/Inputs/Controls

neobrain commented on src/citra_qt/config/inputs.h:
```diff
@@ -0,0 +1,59 @@
+    Settings::Values temp_settings;
+
+    /// Given a key code, return the key name. Needed for modifier keys
+    static QString getKeyName(int key_code);
```
Capital G

neobrain replied:
Don't document who "needs" something, but what it is used for. There's no value in hardcoding assumptions like that into documentation.

neobrain replied:
When returning non-trivial non-reference types, make the return types `const` to prevent accidental misusage as a reference.

chinhodado replied:
> Don't document who "needs" something, but what it is used for. There's no value in hardcoding assumptions like that into documentation.

I just added that explanation for easier understanding why that function is needed. 

> When returning non-trivial non-reference types, make the return types const to prevent accidental misusage as a reference.

I did some Googling and it seems returning const is discouraged?

neobrain replied:
>I just added that explanation for easier understanding why that function is needed. Otherwise people may find it weird that we can't use QKeySequence(key_code).toString(); directly.
You can describe this in a clearer fashion by instead saying "in addition to QKey..., this will translate modifier keys properly".


> I did some Googling and it seems returning const is discouraged?

Not by us.

neobrain commented on src/citra_qt/config/QLineEditKeyConfig.cpp:
```diff
@@ -0,0 +1,100 @@
+
+    int key = event->key();
+    switch (button) {
+    case Button::A:
```
Why not just emit some signal here and pass `button` and `key` as parameters to that? that way, you don't need to store a reference to the config in a utility class (which is kind of an ugly thing to do).

chinhodado replied:
The class is QLineEditKeyConfig, so it's not that unusual to have access to the config. Unless we have some centralized config-changing infrastructure already, I think putting this here is fine.

neobrain replied:
This design is pretty flawed, though. There's no separation of tasks between QLineEditKeyConfig and the input dialog class at all. This clearly shows already in how you define things in this file and use it in inputs.cpp and vice-versa.

Emitting a signal here and moving this switch case to inputs.cpp is a lot clearer, and doesn't require more lines of code either.

turbedi commented on src/citra_qt/config/inputs.h:
```diff
@@ -0,0 +1,59 @@
+    QLineEditKeyConfig *lineEdit_Select;
+    QLineEditKeyConfig *lineEdit_Home;
+
+    /// A vector for all the lineEdits above for easy iterating though them
```
*through

chinhodado commented:
Addressed the comments

neobrain commented on src/citra_qt/config/QLineEditKeyConfig.cpp:
```diff
@@ -0,0 +1,99 @@
+#include "inputs.h"
+#include <QKeyEvent>
+
+std::map<Button, std::string> ButtonNameMap({
```
This will not translate properly. Also, there's not much point in this map, since it's only used exactly once anyway.

chinhodado replied:
How would you replace it then?

neobrain replied:
Remove this map, and at the place it was used before create all labels beforehand.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+    lineEdit_Select = new QLineEditKeyConfig(this, temp_settings, Button::Select);
+
+    lineEdits = {{
+        lineEdit_A, lineEdit_B, lineEdit_X, lineEdit_Y, lineEdit_L, lineEdit_R, lineEdit_ZL, lineEdit_ZR,
```
Why not just initialize these lineEdits directly?

neobrain commented on src/citra_qt/config/inputs.h:
```diff
@@ -0,0 +1,60 @@
+    static QString GetKeyName(int key_code);
+
+private:
+    QLineEditKeyConfig *lineEdit_A;
```
These member variables aren't necessary.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+    temp_settings = Settings::Values(Settings::values);
+
+    // create the lineEdits
+    lineEdit_A = new QLineEditKeyConfig(this, temp_settings, Button::A);
```
This block of code should be simplified to a small loop over `lineEdits`.

neobrain commented on src/citra_qt/config/QLineEditKeyConfig.h:
```diff
@@ -0,0 +1,34 @@
+class QKeyEvent;
+
+/// An enum for the buttons on the 3DS
+enum Button {
```
Are you sure we don't already have such an enum?

chinhodado replied:
Yes I am sure (and this is the nth times you asked this already)

neobrain commented on src/citra_qt/config/QLineEditKeyConfig.cpp:
```diff
@@ -0,0 +1,99 @@
+        temp_settings.pad_zl_key = key;
+        break;
+    case Button::ZR:
+        temp_settings.pad_zr_key = key;
```
Sorry, but this redundancy is really terrible. While it's not your fault, it's not too much effort either to unify the `temp_settings` button members into a single array so that all of these blocks can be simplified to loops. Could you do that, please?

chinhodado replied:
I'm not sure I understand it correctly, but you want to change, in settings.h:

    int pad_a_key;
    int pad_b_key;
    int pad_x_key;
    int pad_y_key;
    ...

into a single array?

That would require quite a lot of work, since those individually values are already used everywhere (bootmanager.cpp, config.cpp, emu_window_glfw.cpp, etc.)

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+}
+
+void GInputsDialog::DisplayButtonSettings(const Settings::Values& values) {
+    lineEdit_A->setText(GetKeyName(values.pad_a_key));
```
This block should be replaced with a simple loop over all line edits (cf. my comment above about changing Settings::values).

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+
+void GInputsDialog::RestoreDefaultSettings() {
+    // load the default button settings into temp_settings
+    temp_settings.pad_a_key      = Qt::Key_A;
```
default key bindings can be stored in an array (or a map), and then this whole block can be replaced by a small loop. (Settings::values needs to be changed first, cf. my comment above.)

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+    Config config;
+
+    // load the temporary settings into our real settings
+    Settings::values.pad_a_key      = temp_settings.pad_a_key;
```
Yet another block of code which should be changed to a small loop.

neobrain commented on src/citra_qt/config/inputs.h:
```diff
@@ -0,0 +1,60 @@
+#include <array>
+#include <QDialog>
+
+#include "../core/settings.h"
```
Don't mix headers like that.

First standard headers, then Qt headers, then our own headers, please.

neobrain commented on src/citra_qt/config/inputs.h:
```diff
@@ -0,0 +1,60 @@
+    GInputsDialog(QWidget* parent = nullptr);
+
+    /// Temporary settings used when configuration is changed but not saved yet
+    Settings::Values temp_settings;
```
It should not be necessary to have this exposed publicly.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+
+    // then save it
+    config.Save();
+    LOG_INFO(Config, "Input settings changed.");
```
I don't think this log message is necessary.

chinhodado replied:
It's just a little confirmation, it doesn't hurt.

neobrain replied:
It does. There's little point in having a log if only useless information is printed to it.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+        int labelColumn = i / numRow * 2;
+        int editColumn = labelColumn + 1;
+        QLineEditKeyConfig *lineEdit = lineEdits[i];
+        QLabel *label = new QLabel(QString::fromStdString(ButtonNameMap[lineEdit->button]), this);
```
there's no need to specify a parent for widgets which are about to be inserted into a layout. The layout takes ownership automatically.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+    for (size_t i = 0; i < size; i++) {
+        int row = i % numRow;
+        int labelColumn = i / numRow * 2;
+        int editColumn = labelColumn + 1;
```
I think `labelColumn+1` is obvious enough that there is no need to store this in a local variable.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+
+    // put the lineEdits and their labels into a grid
+    QGridLayout *grid = new QGridLayout();
+    const int numRow = 8;
```
should be moved into the for loop. Declare variables close to their usage.

chinhodado replied:
It is a const, why would you want to put it into the loop?

neobrain replied:
It's only used inside the loop.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+    }};
+
+    // put the lineEdits and their labels into a grid
+    QGridLayout *grid = new QGridLayout();
```
please don't use () when calling default constructors

Subv replied:
Why?

neobrain replied:
There's no reason to use them in the first place, and we generally don't use them like that anywhere else AFAIK.

chinhodado replied:
I think this is pretty arbitrary. Is there a good reason for it?

(and yes, it is being used like this in a lot of places in the project already)

neobrain replied:
Hmm there's probably no reason to enforce it if it's already used commonly throughout the project.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+    lineEdit_X = new QLineEditKeyConfig(this, temp_settings, Button::X);
+    lineEdit_Y = new QLineEditKeyConfig(this, temp_settings, Button::Y);
+    lineEdit_L = new QLineEditKeyConfig(this, temp_settings, Button::L);
+    lineEdit_R = new QLineEditKeyConfig(this, temp_settings, Button::R);
```
There's no need to specify a parent widget for any of these.

neobrain commented on src/citra_qt/config/QLineEditKeyConfig.h:
```diff
@@ -0,0 +1,34 @@
+public:
+    Button button;
+    Settings::Values& temp_settings;
+    QLineEditKeyConfig(QWidget* parent, Settings::Values& temp_settings, Button button);
```
The parent widget parameter is usually put to the last position and defaults to `nullptr`

neobrain commented on src/citra_qt/config/QLineEditKeyConfig.h:
```diff
@@ -0,0 +1,34 @@
+    Q_OBJECT
+
+public:
+    Button button;
```
these two shouldn't be public.

neobrain commented on src/citra_qt/main.ui:
```diff
@@ -153,6 +154,11 @@
   </action>
+  <action name="action_Inputs">
+   <property name="text">
+    <string>Configure &amp;Controls</string>
```
This has similar context to the "Configure..." option, and hence should be moved next to it.

neobrain commented on src/citra_qt/CMakeLists.txt:
```diff
@@ -24,6 +26,8 @@ set(SRCS
             )
 
 set(HEADERS
+            config/inputs.h
```
These should be ordered alphabetically.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+        grid->addWidget(lineEdit, row, editColumn);
+    }
+
+    // the button box
```
There's no need for comments which just repeat what the source code already obviously says.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+
+    connect(buttonBox, SIGNAL(accepted()), this, SLOT(accept()));
+    connect(buttonBox, SIGNAL(rejected()), this, SLOT(reject()));
+    QMetaObject::connectSlotsByName(this);
```
Please don't use connectSlotsByName. It's confusing and there's only little point at all.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+
+    setLayout(verticalLayout);
+    setWindowTitle(tr("Input Settings"));
+    resize(165, height());
```
Don't manually set sizes. Is this necessary at all?

chinhodado replied:
This just sets the initial width. Otherwise, the text edits would be too wide. You can always resize the dialog at will.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+    DisplayButtonSettings(Settings::values);
+}
+
+void GInputsDialog::DisplayButtonSettings(const Settings::Values& values) {
```
"Display" is a bad term for this, since it doesn't actually display anything. Suggestion: UpdateKeyLabels.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,203 @@
+void GInputsDialog::DisplayButtonSettings(const Settings::Values& values) {
+    lineEdit_A->setText(GetKeyName(values.pad_a_key));
+    lineEdit_B->setText(GetKeyName(values.pad_b_key));
+    lineEdit_X->setText(GetKeyName(values.pad_x_key));
```
Why is values passed in as a parameter instead of just accessing it as a class member?

chinhodado commented:
Address the comments again.

As I said above, I won't make the changes to the Settings::values struct in this PR, since it affects a lot of other unrelated files.

neobrain commented:
Sorry for going nitpicky in my previous review cycles; I should've delayed the details until the important changes are implemented.

You do seem to have addressed the nitpicks, but some important aspects are still not addressed, and I don't think it's sensible to merge this PR in this state. In particular, I suggest blocking this PR for issue #870.

chinhodado commented:
Updated. Moved stuffs into loops thanks to #873 just like @neobrain wants. Should be good now.

neobrain commented on src/citra_qt/config/QLineEditKeyConfig.cpp:
```diff
@@ -0,0 +1,29 @@
+    QString key_name = GInputsDialog::GetKeyName(event->key());
+    setText(key_name);
+
+    ValueChanged(button, event->key());
```
Missing an "emit" here?

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    // the button box that contains the restore default/ok/cancel buttons
+    QVBoxLayout *verticalLayout = new QVBoxLayout(this);
```
Redundant parent specification.

yuriks replied:
Why?

neobrain replied:
ownership is transfered automatically in setLayout().

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,114 @@
+
+    // the button box that contains the restore default/ok/cancel buttons
+    QVBoxLayout *verticalLayout = new QVBoxLayout(this);
+    QDialogButtonBox *buttonBox = new QDialogButtonBox(this);
```
same

chinhodado commented:
Fixed

lioncash commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    // put the lineEdits and their labels into a grid
+    QGridLayout *grid = new QGridLayout();
```
Asterisk goes against the type.

lioncash commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,114 @@
+    QGridLayout *grid = new QGridLayout();
+    size_t size = lineEdits.size();
+    for (size_t i = 0; i < size; i++) {
+        const int numRow = 8;
```
```cpp
static const int NUM_ROW = 8;
```

lioncash commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,114 @@
+        int row = i % numRow;
+        int labelColumn = (int)i / numRow * 2;
+        QLineEditKeyConfig *lineEdit = lineEdits[i];
+        QLabel *label = new QLabel(tr(ButtonNameMap[lineEdit->button].c_str()));
```
Asterisks against type.

lioncash commented on src/citra_qt/config/inputs.h:
```diff
@@ -0,0 +1,38 @@
+#include <QDialog>
+
+#include "../core/settings.h"
+#include "QLineEditKeyConfig.h"
```
This can be made a forward declaration.

lioncash commented on src/citra_qt/config/QLineEditKeyConfig.cpp:
```diff
@@ -0,0 +1,29 @@
+#include "QLineEditKeyConfig.h"
+#include "inputs.h"
+
+using namespace Settings::NativeInput;
```
This isn't really necessary, considering it's only used in two places.

chinhodado replied:
It is used for every button name in the map below.

chinhodado commented:
Done

lioncash commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,115 @@
+    QGridLayout* grid = new QGridLayout();
+    size_t size = lineEdits.size();
+    for (size_t i = 0; i < size; i++) {
+        const int NUM_ROW = 8;
```
This should be static const

yuriks replied:
It doesn't really matter (and I think that for something like this I'd even keep it non-static, less chance of C compiler non-opt surprises.)

lioncash replied:
Fair enough. On another note, I'd be very surprised if a compiler cannot optimize due to a static primitive.

chinhodado commented:
Waiting for @neobrain 's review

neobrain commented:
Sorry, but the new design has lots of usability issues:
* There is no real order in how the button inputs are laid out. In particular, related button input boxes aren't put next to each other. "S-Up" is at the bottom of the second column, while all other "S-" controls are at the top of the third column. The old design had all buttons neatly grouped into framed boxes.
* Those three "columns" are confusing. I can't really put in words why that is, but for what it's worth I don't think I've ever seen such a design in any other application. Consider making the whole thing a list instead. (See also my final remark at the bottom.)
* Use of unclear abbreviations like "C-Up" and "S-Up". I guess the intent was to make the dialog save some horizontal space, but that would only be an issue due to the "three columns" thing mentioned above. The full name could also just be put in a QGroupBox such that the input controls only need to be labelled "Up", "Down", etc.
* Pressing ESC/Enter should exit the dialog (and the configuration changes should be discarded/saved). For instance, consider the case that someone accidentally opens that dialog and just wants to close it again by hitting ESC, or that they changed some key mappings and want to save them by hitting "Enter".
* "Tab" cannot be bound to a key because it cycles to the next input widget. On the other hand, trying to cycle back via "Shift+Tab" will bind "Shift" to the selected input.
* The two points above (ESC/Enter and Tab) seem to stem from the fact that the input boxes grab input immediately instead of after being activated in some way. A solution might be to move from QLineEdits to explicit QPushButtons. Alternatively, I suppose a widget looking like a QLineEdit but which grabs input upon activating it (e.g. via mouse-click) would work too (not sure to what extent a QLineEdit could be configured to do this).
* The button QLabels should end with a double colon
* It always displays blank inputs for me when starting up. No idea why that is.
* Arrow keys and others are not being translated. Not sure if this is simply due to a lack of translation files, or if it's actually a bug. Just noting it, for now.
* Don't hardcode the dialog size. The hardcoded number doesn't scale to system configurations other than yours (case in point: The hardcoded dialog size is below the minimum dialog size on my system anyway, hence making no difference).

By the way,  I realized during review that my old "hotkeys" configuration dialog has been ported to Citra, and I wonder if it might be a better idea to just put the emulator key configuration in there, too (rationale being that from a user perspective there's not much difference between "hotkeys" and "key mappings").

neobrain commented:
After some discussion on IRC, it seems that @archshift and @yuriks (one way or the other) agree with the concern about input grabbing by default, and that the consensus is to change the QLineEdits to QPushButtons, since they don't suffer from any of the mentioned issues.

@chinhodado, I recall having the "QLineEdit vs QPushButton" topic come up shortly on IRC once. Was there actually any particular reason to use QLineEdits in favor of QPushButtons?

chinhodado commented:
> There is no real order in how the button inputs are laid out. In particular, related button input boxes aren't put next to each other. "S-Up" is at the bottom of the second column, while all other "S-" controls are at the top of the third column.

Fixed. It's shown as the screenshot now.

> Use of unclear abbreviations like "C-Up" and "S-Up"

That's the way it is in the code. I have no issue with a more verbose name, but I don't think short names really matter that much.

> tab stuffs

Tab is now captured properly. The QLineEdits are no longer having focus while the dialog is opened initially, and they can't have tab focus.

> The button QLabels should end with a double colon

Done

> It always displays blank inputs for me when starting up.

It was fixed by #1005. I also rebased all the commits on top of master.

> Don't hardcode the dialog size

It just sets the width to make sure the QLineEdits are not too long initially (you can remove that line and see for yourself)

> By the way, I realized during review that my old "hotkeys" configuration dialog has been ported to Citra, and I wonder if it might be a better idea to just put the emulator key configuration in there, too

AFAIK it's not working atm, and I don't want to spend time making it work.

> I recall having the "QLineEdit vs QPushButton" topic come up shortly on IRC once. Was there any particular reason to use QLineEdits in favor of QPushButtons?

I don't really remember having discussed it. No particular reason, just that I like QLineEdits more since it feels more direct. It's true that it has some annoyance with focusing and stuffs, but I don't really mind it that much.

neobrain commented:
I'm going to be brief here to save everybody some time.

> >  There is no real order in how the button inputs are laid out. In particular, related button input boxes aren't put next to each other. "S-Up" is at the bottom of the second column, while all other "S-" controls are at the top of the third column.

> Fixed. The order is as in the screenshot now.

That looks a lot better. I suppose you should still add QGroupBoxes about related controls, though.

> >    Use of unclear abbreviations like "C-Up" and "S-Up"

> That's the way it is in the code. I have no issue with a more verbose name, but I don't think short names really matter that much.

Use the more verbose name. Nothing of what's already in Citra should stop you from doing that.

> >    Don't hardcode the dialog size

> It just sets the initial width to make sure the QLineEdits are not too long initially like this:

Just drop the resize line.



> >    I recall having the "QLineEdit vs QPushButton" topic come up shortly on IRC once. Was there any particular reason to use QLineEdits in favor of QPushButtons?

> I don't really remember having discussed it. No particular reason, just that I like QLineEdits more since it feels more direct. It's true that it has some annoyance with focusing and stuffs, but I don't really mind it that much.

QPushButtons it is, then. A single click would make them grab input (and change text to something indicating that they are waiting for input) and entering a key would make them release input again. Bonus points if you add a timeout after 5 seconds or so.

neobrain commented on src/citra_qt/config/inputs.cpp:
```diff
@@ -0,0 +1,125 @@
+        lineEdits[i]->setFocusPolicy(Qt::ClickFocus);
+    }
+
+    std::string labelArray[] = {
```
These strings are not marked for translation.

chinhodado replied:
They are translated when used [here](https://github.com/citra-emu/citra/pull/531/files#diff-684afa47d52a38683807571e34cfe45bR44)

chinhodado commented:
> Use the more verbose name

What should the verbose names be?

> Just drop the resize line

So you're OK with the layout being wide like that initially?

chinhodado commented:
I give up. It's pretty clear I can't satisfy neobrain no matter what. Anyone wants to pick this up, feel free.
